Added blog post about adding inking canvas to a RNW app (#447)
microsoft/react-native-windows-samples: website/blog/2021-05-25-inking.md

---
title: "Add inking to your React Native for Windows application"
author: Matteo Pagani
authorURL: https://twitter.com/qmatteoq
description: "How to use add inking to a React Native for Windows application"
is_blog: true
published_date: May 25, 2021
---

Windows 10 is one of the most powerful platforms on the market when it comes to supporting digital inking experiences. As such, the Universal Windows Platform comes with a dedicated series of controls that you can add to your application to enable inking experiences. Thanks to these controls, you can easily implements features like capturing signatures, text and shape recognition, etc.

> If you're building a Windows application with React Native, there are a few options you can explore to add such a feature. One of them is using a community module called [react-native-sketch-canvas](https://github.com/creambyemute/react-native-sketch-canvas), which provides a Windows implementation as well. This module is based on Win2D and it enables the recording and playback of ink strokes. While this solution can be a great fit for many applications, there are scenarios in which you need more precision and better integration with the Windows ecosystem, like when you need to capture a signature. For these purposes, Windows 10 offers a native control called [`InkCanvas`](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.InkCanvas),  which enables powerful inking experiences. Since `InkCanvas` is a specific Windows 10 control, it isn't automatically exposed by React Native, since we don't have an equivalent control on other platforms. As such, we'll have to wrap it with a [native UI component](https://microsoft.github.io/react-native-windows/docs/view-managers), which is a feature provided by React Native which exposes native UI controls for Windows, Android or iOS to the JavaScript layer using JSX.

Reviewer: ```suggestion
If you're building a Windows application with React Native, there are a few options you can explore to add such a feature. One of them is using a community module called [react-native-sketch-canvas](https://github.com/creambyemute/react-native-sketch-canvas), which provides a Windows implementation as well. This module is based on Win2D and it enables the recording and playback of ink strokes. While this solution can be a great fit for many applications, there are scenarios in which you need more precision and better integration with the Windows ecosystem, like when you need to capture a signature. For these purposes, Windows 10 offers a native control called [`InkCanvas`](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.InkCanvas), which enables powerful inking experiences. `InkCanvas` is specifically a Windows control and isn't exposed automatically by React Native. As such, we'll have to wrap it with a [native UI component](https://microsoft.github.io/react-native-windows/docs/view-managers), which is a feature provided by React Native which exposes native UI controls for Windows, Android or iOS to the JavaScript layer using JSX.
```

---
title: "Add inking to your React Native for Windows application"
author: Matteo Pagani
authorURL: https://twitter.com/qmatteoq
description: "How to use add inking to a React Native for Windows application"
is_blog: true
published_date: May 25, 2021
---

Windows 10 is one of the most powerful platforms on the market when it comes to supporting digital inking experiences. As such, the Universal Windows Platform comes with a dedicated series of controls that you can add to your application to enable inking experiences. Thanks to these controls, you can easily implements features like capturing signatures, text and shape recognition, etc.

> If you're building a Windows application with React Native, there are a few options you can explore to add such a feature. One of them is using a community module called [react-native-sketch-canvas](https://github.com/creambyemute/react-native-sketch-canvas), which provides a Windows implementation as well. This module is based on Direct2D and it enables the recording and playback of ink strokes. While this solution can be a great fit for many applications, there are scenarios in which you need more precision and better integration with the Windows ecosystem, like when you need to capture a signature. For these purposes, Windows 10 offers a native control called [`InkCanvas`](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.InkCanvas),  which enables powerful inking experiences. Since `InkCanvas` is a specific Windows 10 control, it isn't automatically exposed by React Native, since we don't have an equivalent control on other platforms. As such, we'll have to wrap it with a [native UI component](https://microsoft.github.io/react-native-windows/docs/view-managers), which is a feature provided by React Native which exposes native UI controls for Windows, Android or iOS to the JavaScript layer using JSX.

Reviewer: ```suggestion
If you're building a Windows application with React Native, there are a few options you can explore to add such a feature. One of them is using a community module called [react-native-sketch-canvas](https://github.com/creambyemute/react-native-sketch-canvas), which provides a Windows implementation as well. This module is based on Win2D and it enables the recording and playback of ink strokes. While this solution can be a great fit for many applications, there are scenarios in which you need more precision and better integration with the Windows ecosystem, like when you need to capture a signature. For these purposes, Windows 10 offers a native control called [`InkCanvas`](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.InkCanvas),  which enables powerful inking experiences. Since `InkCanvas` is a specific Windows 10 control, it isn't automatically exposed by React Native, since we don't have an equivalent control on other platforms. As such, we'll have to wrap it with a [native UI component](https://microsoft.github.io/react-native-windows/docs/view-managers), which is a feature provided by React Native which exposes native UI controls for Windows, Android or iOS to the JavaScript layer using JSX.
```

---
title: "Add inking to your React Native for Windows application"
author: Matteo Pagani
authorURL: https://twitter.com/qmatteoq
description: "How to use add inking to a React Native for Windows application"
is_blog: true
published_date: May 25, 2021
---

Windows 10 is one of the most powerful platforms on the market when it comes to supporting digital inking experiences. As such, the Universal Windows Platform comes with a dedicated series of controls that you can add to your application to enable inking experiences. Thanks to these controls, you can easily implements features like capturing signatures, text and shape recognition, etc.

> If you're building a Windows application with React Native, there are a few options you can explore to add such a feature. One of them is using a community module called [react-native-sketch-canvas](https://github.com/creambyemute/react-native-sketch-canvas), which provides a Windows implementation as well. This module is based on Direct2D and it enables the recording and playback of ink strokes. However, if this solution can be a great fit for many applications, there are scenarios in which you need more precision and better integration with the Windows ecosystem, like when you need to capture a signature. For these purposes, Windows 10 offers a native control called `InkCanvas`,  which enables powerful inking experiences. However, being a specific Windows 10 control, it isn't automatically exposed by React Native, since we don't have an equivalent control on other platforms. As such, we'll have to wrap it with a [native UI component](https://microsoft.github.io/react-native-windows/docs/view-managers), which is a feature provided by React Native which enables to expose native UI controls for Windows, Android or iOS to the JavaScript layer using JSX.

Reviewer: ```suggestion
If you're building a Windows application with React Native, there are a few options you can explore to add such a feature. One of them is using a community module called [react-native-sketch-canvas](https://github.com/creambyemute/react-native-sketch-canvas), which provides a Windows implementation as well. This module is based on Direct2D and it enables the recording and playback of ink strokes. While this solution can be a great fit for many applications, there are scenarios in which you need more precision and better integration with the Windows ecosystem, like when you need to capture a signature. For these purposes, Windows 10 offers a native control called [`InkCanvas`](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.InkCanvas),  which enables powerful inking experiences. Since `InkCanvas` is a specific Windows 10 control, it isn't automatically exposed by React Native, since we don't have an equivalent control on other platforms. As such, we'll have to wrap it with a [native UI component](https://microsoft.github.io/react-native-windows/docs/view-managers), which is a feature provided by React Native which exposes native UI controls for Windows, Android or iOS to the JavaScript layer using JSX.
```

> In this article we'll walk through the process of wrapping the `InkCanvas` controls, with the dual goal of enabling your apps to support inking and walking you through the process of creating a native UI component, which can be useful also for other controls.

Reviewer: ```suggestion
In this article we'll walk through the process of wrapping the `InkCanvas` control, with the dual goals of enabling your apps to support inking and walking you through the process of creating a native UI component, which can be useful for exposing other native controls.
```

---
title: "Add inking to your React Native for Windows application"
author: Matteo Pagani
authorURL: https://twitter.com/qmatteoq
description: "How to use add inking to a React Native for Windows application"
is_blog: true
published_date: May 25, 2021
---

Windows 10 is one of the most powerful platforms on the market when it comes to supporting digital inking experiences. As such, the Universal Windows Platform comes with a dedicated series of controls that you can add to your application to enable inking experiences. Thanks to these controls, you can easily implements features like capturing signatures, text and shape recognition, etc.

> The starting point is the `InkCanvas` control which provides a canvas that is able to react to inking interactions. However, being a specific Windows 10 control, it isn't automatically exposed by React Native, since we don't have an equivalent control on other platforms. As such, we'll have to wrap it with a [native UI component](https://microsoft.github.io/react-native-windows/docs/view-managers), which is a feature provided by React Native which enables to expose native UI controls for Windows, Android or iOS to the JavaScript layer using JSX. 

Reviewer 2: ```suggestion
The starting point is the XAML [`InkCanvas`](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.InkCanvas) control which provides a canvas that handles inking interactions. However, being a native XAML control, it isn't automatically exposed by React Native, since we don't have an equivalent control on other platforms. As such, we'll have to wrap it with a [native UI component](https://microsoft.github.io/react-native-windows/docs/view-managers), which is a feature provided by React Native which enables to expose native UI controls for Windows, Android or iOS to the JavaScript layer using JSX. 
```

The method receives, as input, a reference to the user control and the value assigned via JSX. The implementation of the method is straigthforward: if the property assigned in JSX has a value, we assign it to the `Title` property of the user control; otherwise, we clear it.

Now we'll be able to leverage the `title` property in our React Native components, like in the following sample:

```javascript
class App extends React.Component {

  render(){
    return (
      <View>
        <InkCanvas style={{ width: 1000, height: 1000}}
                   title="Hello world!" />
      </View>
    );
  }
};
```

And this is the final result:

![](assets/2021-05-25-inking/HelloWorld.png)

### Exposing a command

One of the most common features that developers want to implement when they work with inking is the ability to save the content of the `InkCanvas` control as an image. To enable this feature we're going to implement a command. The concept is very similar to exposing a method in a React Native module and it will enable us to trigger, from JavaScript, the native APIs exposed by the `InkCanvas` control to render the ink into a stream.

> The first step is to install a library called [Win2D](https://github.com/microsoft/Win2D), which enables advanced 2D operations in a UWP app. Thanks to this library, we'll be able to render the content of the canvas into a stream that we can save into a file. To install it, right click on your Windows Runtime Component, choose **Manage NuGet Packages** and look for the [Win2D package](https://www.nuget.org/packages/Win2D.uwp/).

Reviewer 2: curious, is this necessary at all? seems like one should be able to save the content of the ink canvas without using win2D (which is also C# only afaik)

Author: Let me investigate and see if there's another way. Last time I checked, not using Win2D means basically reimplementing what Win2D does to render the canvas. Unfortunately the control doesn't expose any straightforward API to save the content into an image.
